### ISSUE TITLE
[clang-tidy][false-positive] `readability-redundant-typename` suggests syntax error in `release/22.x`

### ISSUE THREAD

**Issue by tearfur** [llvm/llvm-project#206995]
Breaking commit: 80f11cb923ee8551f673ebf59a2ee64dea261fba

In `release/22.x`, `readability-redundant-typename` reports some false positives and the suggested fix is a syntax error.

No godbolt reproducer because this is not in `main`.

```console
$ clang-tidy -checks=readability-redundant-typename typename.cpp -- -std=c++20
1 warning generated.
/tmp/typename.cpp:15:42: warning: redundant 'typename' [readability-redundant-typename]
   15 |     !std::is_same_v<C, std::basic_string<typename C::value_type>>;
      |                                          ^~~~~~~~
```

```cpp
// typename.cpp
#include <string>
#include <type_traits>
#include <utility>
#include <vector>

template<typename C, typename = void>
inline constexpr bool a = false;

template<typename C>
inline constexpr bool a<
    C,
    std::void_t<
        typename C::value_type,
        decltype(std::declval<C&>().push_back(std::declval<typename C::value_type const&>()))>> =
    !std::is_same_v<C, std::basic_string<typename C::value_type>>;

static_assert(a<std::vector<int>>);
```

**Comment by vbvictor** [comment]
If it isn't on main we can't really do about it.
Latest 22.x release was due yesterday per https://llvm.org/.
So the only way is to wait for 23.x which is scheduled `Tue Aug 25th, 2026`.
Or the earliest possible `Tue Jul 14th, 2026: release/23.x branch`

**Comment by vbvictor** [comment]
Confurmed no repro https://godbolt.org/z/4E83rTv76

**Comment by tearfur** [comment]
Understood that this won't be fixed in 22.x.

Still, since the current tests were not able to catch this regression, I suggest adding the reproducer to tests on `main`.

**Comment by zeyi2** [comment]
> Understood that this won't be fixed in 22.x.
> 
> Still, since the current tests were not able to catch this regression, I suggest adding the reproducer to tests on `main`.

I opened https://github.com/llvm/llvm-project/pull/207318 for it, thanks!

**Comment by tearfur** [comment]
Thank you!
